### PR TITLE
feat: allow range with two arguments and bound

### LINKED ISSUE #3678: allow `range(x, y, bound=N)`
per title. `bound=` is only allowed in the single arg form of `range()`, but it should be possible for this one too

ex.
```vyper
for i in range(1, N_COINS, bound=MAX_COINS):
  ...
```

### PR DESCRIPTION
**Pull request by DanielSchiavini**
Fixes #3678

### What I did
- Add support for `range` with 2 arguments
- Adds test coverage for touched code
- Add new range format to documentation

### How I did it
- Updated codegen to utilize the new arguments
- Updated semantic analysis to handle new cases

### How to verify it
Tests are included

### Commit message

`Allow range(x, y, bound=N)`

### Description for the changelog
Add support for range with start/end runtime arguments as long as a `bound` argument is given.

### Cute Animal Picture
image

### CODE REVIEW

File: vyper/codegen/stmt.py [hunk 1]
```diff
@@ -277,14 +278,21 @@ def _parse_For_range(self):
...
             start = Expr.parse_value_expr(arg0, self.context)
             _, hi = start.typ.int_bounds
             start = clamp("le", start, hi + 1 - rounds)
             rounds_bound = rounds
 
+        # Type 4 for, e.g. for i in range(1, x, bound=4): ...
+        else:
+            start = Expr.parse_value_expr(arg0, self.context)
+            end = Expr.parse_value_expr(self.stmt.iter.args[1], self.context)
+            rounds = IRnode.from_list(["sub", end, start], typ=iter_typ)
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:293]: i think an underflow check is needed here

File: vyper/semantics/analysis/local.py [hunk 2]
```diff
@@ -347,66 +347,62 @@ def visit_For(self, node):
                 )
             range_ = node.iter
             validate_call_args(range_, (1, 2), kwargs=["bound"])
+            bound = None
+            if range_.keywords:
+                bound = range_.keywords[0].value  # this is the only kwarg accepted by validation
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:352]: i'd prefer to leave the old way (constructing a dict out of kwargs) because it scales better

File: vyper/codegen/stmt.py (hunk 1 above)
Comment (DanielSchiavini) [reply vyper/codegen/stmt.py:293]: I am trying to reproduce an underflow in tests. However [this test](https://github.com/DanielSchiavini/vyper/blob/a6fb6a059ebda8403044de05b36ca925cb664858/tests/functional/codegen/features/iteration/test_for_range.py#L68) seems to cover the possibility of underflow. Am I missing something?

File: vyper/semantics/analysis/local.py (hunk 2 above)
Comment (DanielSchiavini) [reply vyper/semantics/analysis/local.py]: Fixed

File: vyper/codegen/stmt.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/codegen/stmt.py:293]: yea i think that happens to be caught because of the range check in `repeat`. `-1` in two's complement is much greater than 6. but you can construct a start+end/bound which will pass the test, maybe something like `end - start == -5` and `bound = 4`

Comment (DanielSchiavini) [reply vyper/codegen/stmt.py]: I could not reproduce the underflow, but I've added a runtime check

File: vyper/codegen/stmt.py [hunk 3]
```diff
@@ -254,37 +254,43 @@ def _parse_For_range(self):
...
             start = Expr.parse_value_expr(arg0, self.context)
             _, hi = start.typ.int_bounds
             start = clamp("le", start, hi + 1 - rounds)
             rounds_bound = rounds
 
+        # Type 4 for, e.g. for i in range(x, y, bound=N): ...
+        else:
+            start = Expr.parse_value_expr(arg0, self.context)
+            end = Expr.parse_value_expr(arg1, self.context)
+            rounds = IRnode.from_list(["sub", end, clamp("le", start, end)])
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:291]: for re-used variables we currently need to use `with end.cache_with_complex("end") as (b1, end):` since that handles side effects

Comment (charles-cooper) [reply vyper/codegen/stmt.py:291]: you can see this with something like
```
@internal
def some_side_effects() -> uint256:  # for illustration
    self.counter += 1
    return 10

@external
def foo():
    for i in range(1, self.some_side_effects(), bound=100):
        ...
    assert self.counter == 1
```

Comment (DanielSchiavini) [reply vyper/codegen/stmt.py]: This example raises `StateAccessViolation: May not call state modifying function 'some_side_effects' within a range expression.`
I could not find a consistent way to test the side effect, but I've added the caching in [Cache the range end when complex](https://github.com/vyperlang/vyper/pull/3679/commits/4aa953f16a1137f0bfb3001f7bec51f43b4689b1)

Comment (charles-cooper) [reply vyper/codegen/stmt.py]: in the new IR it's no longer necessary anyways -- but for now, it's at least good hygiene.

File: vyper/semantics/analysis/local.py [hunk 4]
```diff
@@ -479,8 +415,8 @@ def visit_For(self, node):
 
                 try:
                     with NodeMetadata.enter_typechecker_speculation():
-                        for n in node.body:
-                            self.visit(n)
+                        for start in node.body:
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:418]: not sure `start` is a good name here - how about `stmt`

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 5]
```diff
@@ -31,6 +31,71 @@ def repeat(n: uint256) -> uint256:
     assert_tx_failed(lambda: c.repeat(7))
 
 
+def test_range_bound_two_args(get_contract, assert_tx_failed):
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:34]: maybe we should add a test where start is a variable but end is a constant, ex. `range(start, 100, bound=5)`

**Review by charles-cooper — COMMENTED** [review]
looks nice! nice refactoring of that range functionality too. left a couple comments, i think we also want to clarify the behavior of the bounds check in the docs (i.e., `end - start <= bound` as opposed to, for instance, `end <= bound`).

File: vyper/semantics/analysis/local.py (hunk 4 above)
Comment (charles-cooper) [reply vyper/semantics/analysis/local.py:418]: > How about statement or simply node? Im not sure why I changed this

because they are all of type `vy_ast.Stmt`! re. terminology -- i think `stmt` is the standard abbreviation here (in the codebase, and also the terminology used in the python ast)

File: tests/functional/codegen/features/iteration/test_for_range.py (hunk 5 above)
Comment (DanielSchiavini) [reply tests/functional/codegen/features/iteration/test_for_range.py:34]: To be honest that is not handled in the code at this moment. I based off the existing documentation that specifies this is not supported. I'd be happy to implement this, but maybe a separate PR is better?

Comment (DanielSchiavini) [reply tests/functional/codegen/features/iteration/test_for_range.py:34]: Oh no :scream: this gives a bad error message
```
@external
def repeat(n: uint256) -> uint256:
    x: uint256 = 0
    for i in range(n, 6, bound=6):
        x += i + 1
    return x
```
Fails compilation with `AssertionError: Number of arguments mismatched: calldataload []`.
This definitely needs improvement, even if unsupported.

Comment (DanielSchiavini) [reply tests/functional/codegen/features/iteration/test_for_range.py:52]: I've added support and tests for the error messages :+1:

File: docs/control-structures.rst [hunk 6]
```diff
@@ -293,3 +293,12 @@ Here, ``START`` and ``STOP`` are literal integers, with ``STOP`` being a greater
...
 ``a`` is a variable with an integer type and ``N`` is a literal integer greater than zero.  ``i`` begins as ``a`` and increments by one until it is equal to ``a + N``. If ``a + N`` would overflow, execution will revert.
+
+
+Finally, it is possible to use ``range`` with runtime `start` and `stop` values as long as a constant `bound` value is provided.
+In this case, Vyper guarantees that `y - x <= bound`.
+
+.. code-block:: python
+
+    for i in range(x, y, bound=CONSTANT):
+        ...
```
Comment (charles-cooper) [inline docs/control-structures.rst:304]:
```suggestion
In this case, Vyper checks at runtime that `end - start <= bound`.

.. code-block:: python

    for i in range(start, end, bound=CONSTANT):
        ...
```

File: vyper/semantics/analysis/local.py [hunk 7]
```diff
@@ -790,3 +726,107 @@ def visit_IfExp(self, node: vy_ast.IfExp, typ: VyperType) -> None:
...
+    kwargs = {s.arg: s.value for s in node.keywords or []}
+
+    start, end = (vy_ast.Int(value=0), node.args[0]) if len(node.args) == 1 else node.args
+
+    if "bound" in kwargs:
+        return _analyse_range_bound(start, end, kwargs["bound"], node)
+    if isinstance(start, vy_ast.Num):
+        return _analyse_range_constant(start, end, node)
+    return _analyse_range_sum(start, end, node)
+
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:747]:
```suggestion
    if "bound" in kwargs:
        return _analyse_range_bound(start, end, kwargs["bound"], node)

    if isinstance(start, vy_ast.Num):
        return _analyse_range_constant(start, end, node)

    return _analyse_range_sum(start, end, node)

```

File: vyper/semantics/analysis/local.py [hunk 8]
```diff
@@ -790,3 +726,107 @@ def visit_IfExp(self, node: vy_ast.IfExp, typ: VyperType) -> None:
...
+    :param end: second argument to range()
+    :param bound: bound keyword argument to range()
+    :param node: range() call node
+    """
+    if not isinstance(bound, vy_ast.Num):
+        raise StateAccessViolation("Bound must be a literal", bound)
+    if bound.value <= 0:
+        raise StructureException("Bound must be at least 1", bound)
+    if (
+        isinstance(start, vy_ast.Num)
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:781]: not necessary for this PR but it might be nice if Num nodes had some function like `int_val()` -- this would make it so we don't have to do `isinstance` checks (e.g. `isinstance(bound, vy_ast.Int)` or `isinstance(start.value, int)`

Comment (charles-cooper) [reply vyper/semantics/analysis/local.py:781]: here's an example of triggering a strange error message:

Screenshot from 2023-12-13 07-58-55

(you can't really get a compiler panic or anything because of the check to `get_common_types()` later)

File: tests/functional/syntax/test_for_range.py [hunk 9]
```diff
@@ -138,27 +161,35 @@ def bar():
...
+]
+
 
-@pytest.mark.parametrize("bad_code,expected_error", fail_list)
+@pytest.mark.parametrize("bad_code,expected_error", fail_list, ids=fail_test_names)
 def test_range_fail(bad_code, expected_error):
     with pytest.raises(type(expected_error)) as exc_info:
         compiler.compile_code(bad_code)
     assert expected_error.message == exc_info.value.message
+    assert expected_error.args[1] == exc_info.value.args[1].node_source_code
```
Comment (charles-cooper) [inline tests/functional/syntax/test_for_range.py:192]: this pattern is a bit weird since the second argument to a `VyperException` should be an ast node. maybe just make the expected node_source_code the third item in the tuple?

**Review by charles-cooper — COMMENTED** [review]
looks good! left a small comments and then this PR should be good to go!

File: tests/functional/syntax/test_for_range.py (hunk 9 above)
Comment (DanielSchiavini) [reply tests/functional/syntax/test_for_range.py]: Yeah I decided to [split the message as well](https://github.com/vyperlang/vyper/pull/3679/commits/7284de2a560dfe74a0c4bf78dc1d831101a5f2ff), as the way the exception is constructed is different

File: vyper/semantics/analysis/local.py (hunk 8 above)
Comment (DanielSchiavini) [reply vyper/semantics/analysis/local.py:783]: yeah I think most isinstance calls can be avoided by using properties and duck typing
We can easily improve the error message here if you prefer

Comment (charles-cooper) [reply vyper/semantics/analysis/local.py:783]: i'm going to declare the `int_val()` thing as "out-of-scope" here :)

however, it did just occur to me that maybe for this particular error message, we should just outright ban the use of `bound=` when the arguments are literals?

**Comment by DanielSchiavini** [comment]
@charles-cooper a follow-up PR is ready in https://github.com/DanielSchiavini/vyper/pull/1 to deprecate `range(x, x+N)`

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 10]
```diff
@@ -228,39 +228,13 @@ def foo(a: {typ}) -> {typ}:
...
-    found = False
-    x = max_value({typ}) - 1
-    for i in range(x, x + 2):
-        if i == max_value({typ}):
-            found = True
-
-    assert found
-    """
-    c = get_contract(code)
-    c.test()
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:171]: are we sure we should remove this test? is it superseded by some other test?

File: vyper/codegen/stmt.py [hunk 11]
```diff
@@ -242,36 +242,28 @@ def _parse_For_range(self):
...
+        if args_len == 1:
+            arg0, arg1 = (IRnode.from_list(0, typ=iter_typ), for_iter.args[0])
+        elif args_len == 2:
+            arg0, arg1 = for_iter.args
         else:
-            arg1 = self.stmt.iter.args[1]
-            rounds = self._get_range_const_value(arg1.right)
+            raise InvalidOperation("Invalid number of arguments to range()")
+
+        kwargs = {s.arg: self._get_range_const_value(s.value) for s in for_iter.keywords}
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:254]: let's have `kwargs = {s.arg: s.value for s in for_iter.keywords}` here and then actually compile it later

File: vyper/codegen/stmt.py [hunk 12]
```diff
@@ -242,36 +242,28 @@ def _parse_For_range(self):
...
+
+        if "bound" in kwargs:
             start = Expr.parse_value_expr(arg0, self.context)
-            _, hi = start.typ.int_bounds
-            start = clamp("le", start, hi + 1 - rounds)
+            end = Expr.parse_value_expr(arg1, self.context)
+            with end.cache_when_complex("end") as (builder, end):
+                rounds = builder.resolve(IRnode.from_list(["sub", end, clamp("le", start, end)]))
+            rounds_bound = kwargs["bound"]
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:261]:
```suggestion
            rounds_bound = self._get_range_const_value(kwargs["bound"])
```

File: vyper/codegen/stmt.py [hunk 13]
```diff
@@ -284,10 +276,12 @@ def _parse_For_range(self):
...
-        loop_body = ["seq"]
-        # store the current value of i so it is accessible to userland
-        loop_body.append(["mstore", iptr, i])
-        loop_body.append(parse_body(self.stmt.body, self.context))
+        loop_body = [
+            "seq",
+            # store the current value of i, so it is accessible to userland
+            ["mstore", iptr, i],
+            parse_body(self.stmt.body, self.context),
+        ]
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:284]: let's revert this change to the previous style please, this style ends up with too much nesting

File: vyper/semantics/analysis/local.py [hunk 14]
```diff
@@ -801,3 +736,82 @@ def visit_IfExp(self, node: vy_ast.IfExp, typ: VyperType) -> None:
...
+    start: vy_ast.Num, end: vy_ast.VyperNode, node: vy_ast.Call
+) -> list[VyperType]:
+    """
+    Check that the arguments to a range(N, M) call are valid.
+    :param start: first argument to range()
+    :param end: second argument to range()
+    :param node: range() call node
+    :return: The common types of the arguments
+    """
+    for arg in [start, end]:
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:768]:
```suggestion
    for arg in (start, end):
```

File: vyper/semantics/analysis/local.py [hunk 15]
```diff
@@ -801,3 +736,82 @@ def visit_IfExp(self, node: vy_ast.IfExp, typ: VyperType) -> None:
...
+    """
+    validate_call_args(node, (1, 2), kwargs=["bound"])
+    kwargs = {s.arg: s.value for s in node.keywords or []}
+
+    start, end = (vy_ast.Int(value=0), node.args[0]) if len(node.args) == 1 else node.args
+
+    if "bound" in kwargs:
+        return _analyse_range_bound(start, end, kwargs["bound"], node)
+
+    return _analyse_range_constant(start, end, node)
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:755]: i think the code paths for `analyse_range_bound()` and `analyse_range_constant()` can somehow be merged together now

File: tests/functional/codegen/features/iteration/test_for_range.py (hunk 10 above)
Comment (DanielSchiavini) [reply tests/functional/codegen/features/iteration/test_for_range.py:171]: Without the special handling for the `range(x, x+N)`, `max + 1` will overflow.
Having a range that goes until the edge is not generally useful, since arrays are 0-indexed.
In order to not overflow, the `i` must be calculated later on
```
    for i in range(x - 1, x):
        if i + 1 == max_value({typ}):
            found = True
```

File: vyper/codegen/stmt.py (hunk 13 above)
Comment (DanielSchiavini) [reply vyper/codegen/stmt.py:284]: I do like to strive for immutability, but I'm fine with reverting this

File: vyper/semantics/analysis/local.py [hunk 16]
```diff
@@ -746,72 +746,38 @@ def _analyse_range_call(node: vy_ast.Call) -> list[VyperType]:
...
+        if (
+            isinstance(start, vy_ast.Num)
+            and isinstance(end, vy_ast.Num)
+            and end.value - start.value > bound.value
+        ):
+            raise StructureException(
+                f"For loop has invalid number of iterations ({end.value - start.value}), "
+                f"the value must be between zero and the bound",
+                node,
+            )
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:774]: i'm wondering if we really want to allow this case (`range(N, M, bound=K)` where `N`, `M` and `K` are all constants). i think maybe we should just reject it?

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 17]
```diff
@@ -145,26 +228,28 @@ def foo(a: {typ}) -> {typ}:
     assert c.foo(100) == 31337
 
 
-# test that we can get to the upper range of an integer
 @pytest.mark.parametrize("typ", ["uint8", "int128", "uint256"])
 def test_for_range_edge(get_contract, typ):
+    """
+    Check that we can get to the upper range of an integer.
+    Note that to avoid overflow, given the semantics of range(),
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:235]: hmm is this a problem with the new implementation of `range(x, y, bound=)`? like we can't get to `max_value()` for a given type?

File: vyper/semantics/analysis/local.py (hunk 16 above)
Comment (DanielSchiavini) [reply vyper/semantics/analysis/local.py:774]: Sure, I can remove it

File: tests/functional/codegen/features/iteration/test_for_range.py (hunk 17 above)
Comment (DanielSchiavini) [reply tests/functional/codegen/features/iteration/test_for_range.py:235]: It was not possible to use the format `range(x, y, bound=N)` but only `range(x, x + 1)`. That was because we totally ignored the `x+1` and only looked at the `+1` part:
https://github.com/vyperlang/vyper/blob/c6f457a73db40e4b113497883bd330e0dcec28d1/vyper/codegen/stmt.py#L270-L271

max_value + 1 would otherwise always overflow

Comment (charles-cooper) [reply tests/functional/codegen/features/iteration/test_for_range.py:235]: ah, i see.

File: tests/functional/syntax/test_for_range.py [hunk 18]
```diff
@@ -37,9 +37,8 @@ def foo():
         pass
     """,
         StructureException,
-        "For loop has invalid number of iterations (10), "
-        "the value must be between zero and the bound",
-        "range(10, 20, bound=5)",
+        "Please remove the range argument when using range with constants",
```
Comment (charles-cooper) [inline tests/functional/syntax/test_for_range.py:40]:
```suggestion
        "Please remove the `bound=` kwarg when using range with constants",
```

File: vyper/semantics/analysis/local.py [hunk 19]
```diff
@@ -762,16 +762,9 @@ def _analyse_range_call(node: vy_ast.Call) -> list[VyperType]:
...
-            and isinstance(end, vy_ast.Num)
-            and end.value - start.value > bound.value
-        ):
-            raise StructureException(
-                f"For loop has invalid number of iterations ({end.value - start.value}), "
-                f"the value must be between zero and the bound",
-                node,
-            )
+        if isinstance(start, vy_ast.Num) and isinstance(end, vy_ast.Num):
+            error = "Please remove the range argument when using range with constants"
```
Comment (charles-cooper) [inline vyper/semantics/analysis/local.py:766]:
```suggestion
            error = "Please remove the `bound=` kwarg when using range with constants"
```

File: tests/functional/syntax/exceptions/test_invalid_literal_exception.py [hunk 20]
```diff
@@ -18,13 +18,6 @@ def foo():
...
     """
 @external
-def foo(x: int128):
-    y: int128 = 7
-    for i in range(x, x + y):
-        pass
-    """,
-    """
-@external
 def foo():
```
Comment (charles-cooper) [inline tests/functional/syntax/exceptions/test_invalid_literal_exception.py:21]: does this raise a StructureException now?

File: tests/functional/codegen/features/iteration/test_for_range.py (hunk 17 above)
Comment (charles-cooper) [reply tests/functional/codegen/features/iteration/test_for_range.py:235]:
```suggestion
    Note that to avoid overflow in the bounds check for range(),
```

File: tests/functional/syntax/exceptions/test_invalid_literal_exception.py (hunk 20 above)
Comment (DanielSchiavini) [reply tests/functional/syntax/exceptions/test_invalid_literal_exception.py:21]: yeah this test was duplicated in other files

Comment (charles-cooper) [reply tests/functional/syntax/exceptions/test_invalid_literal_exception.py:21]: hmm i did not actually see this one in any other files, can you point me to it?

**Comment by charles-cooper** [comment]
tests need to be updated to use the new `tx_failed()` fixture!

File: tests/functional/syntax/exceptions/test_invalid_literal_exception.py (hunk 20 above)
Comment (DanielSchiavini) [reply tests/functional/syntax/exceptions/test_invalid_literal_exception.py:21]: There's a lot of tests resulting in "Value must be a literal integer" in [test_for_range]. It already fails on validating the `x` and `x+y` isn't even looked at.
It's not exactly the same code, if you wish we can add this exact code to the tests

**Comment by DanielSchiavini** [comment]
@charles-cooper I can update when we merge it, otherwise the PRs will be dependent on each other

File: tests/functional/syntax/exceptions/test_invalid_literal_exception.py (hunk 20 above)
Comment (DanielSchiavini) [reply tests/functional/syntax/exceptions/test_invalid_literal_exception.py:21]: Sorry here is the link
https://github.com/DanielSchiavini/vyper/blob/d161895/tests/functional/syntax/test_for_range.py#L79-L124

Comment (DanielSchiavini) [reply tests/functional/syntax/exceptions/test_invalid_literal_exception.py:21]: either way, extra test can't hurt
https://github.com/vyperlang/vyper/pull/3679/commits/548ce58d1d875784310be0039b3784ecde70a8ad

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 21]
```diff
@@ -32,6 +32,100 @@ def repeat(n: uint256) -> uint256:
...
+    for i in range(start, end, bound=6):
+        x = i
+    return x
+    """
+    c = get_contract(code)
+    UINT_MAX = 2**256 - 1
+    for n in range(6):
+        expected = UINT_MAX - 1 if n > 0 else 0
+        assert c.get_last(UINT_MAX - n, UINT_MAX) == expected
+
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:120]: i think it would be good to separate out the `0` case here
```suggestion
    assert c.get_last(UINT_MAX, UINT_MAX) == 0  # initial value of x

    for n in range(1, 6):
        assert c.get_last(UINT_MAX - n, UINT_MAX) == UINT_MAX - 1

```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 22]
```diff
@@ -32,6 +32,100 @@ def repeat(n: uint256) -> uint256:
...
+    # check assertion for start >= end
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:121]:
```suggestion
    # check for `start + bound <= end`, overflow cases
```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 23]
```diff
@@ -89,7 +184,7 @@ def test_offset_repeater_2(get_contract_with_gas_estimation, typ):
 @external
 def sum(frm: {typ}, to: {typ}) -> {typ}:
     out: {typ} = 0
-    for i in range(frm, frm + 101):
+    for i in range(frm, frm + 101, bound=101):
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:187]: just as a note, a lot of these can be refactored now. like this could be rewritten as
```vyper
for i in range(frm, to, bound=101):
    ...
```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 24]
```diff
@@ -32,6 +32,101 @@ def repeat(n: uint256) -> uint256:
...
+
+    # check assertion for start >= end
+    with tx_failed():
+        c.repeat(1, 0)
+    with tx_failed():
+        c.repeat(7, 0)
+    with tx_failed():
+        c.repeat(8, 7)
+
+    # check codegen inserts assertion for n greater than bound
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:99]:
```suggestion
    # check codegen inserts assertion for `start + bound > end`
```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 25]
```diff
@@ -32,6 +32,101 @@ def repeat(n: uint256) -> uint256:
...
+    x: uint256 = 0
+    for i in range(1, n, bound=6):
+        x += i + 1
+    return x
+    """
+    c = get_contract(code)
+    for n in range(1, 8):
+        assert c.repeat(n) == sum(i + 1 for i in range(1, n))
+
+    # check assertion for n less than start
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:68]:
```suggestion
    # check assertion for `start <= end`
```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 26]
```diff
@@ -32,6 +32,101 @@ def repeat(n: uint256) -> uint256:
...
+    with tx_failed():
+        c.repeat(0)
+
+    # check codegen inserts assertion for n greater than bound
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:72]:
```suggestion
    # check codegen inserts assertion for `start + bound <= end`
```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 27]
```diff
@@ -32,6 +32,101 @@ def repeat(n: uint256) -> uint256:
...
+    for i in range(start, end, bound=6):
+        x += i
+    return x
+    """
+    c = get_contract(code)
+    for n in range(0, 7):
+        assert c.repeat(0, n) == sum(range(0, n))
+        assert c.repeat(n, n * 2) == sum(range(n, n * 2))
+
+    # check assertion for start >= end
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:91]:
```suggestion
    # check assertion for `start <= end`
```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 28]
```diff
@@ -32,6 +32,101 @@ def repeat(n: uint256) -> uint256:
...
+    x: uint256 = 0
+    for i in range(n, 7, bound=6):
+        x += i + 1
+    return x
+    """
+    c = get_contract(code)
+    for n in range(1, 5):
+        assert c.repeat(n) == sum(i + 1 for i in range(n, 7))
+
+    # check codegen inserts assertion for n greater than bound
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:48]:
```suggestion
    # check codegen inserts assertion for `end - start <= bound`
```

File: tests/functional/codegen/features/iteration/test_for_range.py [hunk 29]
```diff
@@ -32,6 +32,101 @@ def repeat(n: uint256) -> uint256:
...
+    """
+    c = get_contract(code)
+    for n in range(1, 5):
+        assert c.repeat(n) == sum(i + 1 for i in range(n, 7))
+
+    # check codegen inserts assertion for `end - start <= bound`
+    with tx_failed():
+        c.repeat(8)
+    with tx_failed():
+        c.repeat(0)
```
Comment (charles-cooper) [inline tests/functional/codegen/features/iteration/test_for_range.py:52]:
```suggestion
    # check assertion for `start <= end`
    with tx_failed():
        c.repeat(8)
    # check assertion for `start + bound <= end`
    with tx_failed():
        c.repeat(0)
```

File: vyper/codegen/stmt.py [hunk 30]
```diff
@@ -249,36 +245,28 @@ def _parse_For_range(self):
...
+        for_iter: vy_ast.Call = self.stmt.iter
+        args_len = len(for_iter.args)
+        if args_len == 1:
+            arg0, arg1 = (IRnode.from_list(0, typ=iter_typ), for_iter.args[0])
+        elif args_len == 2:
+            arg0, arg1 = for_iter.args
         else:
-            arg1 = self.stmt.iter.args[1]
-            rounds = self._get_range_const_value(arg1.right)
+            raise InvalidOperation("Invalid number of arguments to range()")
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:255]: this should be a panic really

**Review by charles-cooper — APPROVED** [review]
thanks!
